src/test/lib/DataTest.cpp

```diff
@@ -1363,6 +1363,285 @@ QuicAbortiveTransfers(
+    _In_ struct MsQuicStream* /* Stream */,
+    _In_opt_ void* Context,
+    _Inout_ QUIC_STREAM_EVENT* Event
+)
+{
+    if (Context == nullptr) {
+        return QUIC_STATUS_INVALID_PARAMETER;
+    }
+
+    auto TestContext = reinterpret_cast<CancelOnLossContext*>(Context);
+
+    QUIC_STATUS Status = QUIC_STATUS_SUCCESS;
+
+    switch (Event->Type) {
+    case QUIC_STREAM_EVENT_RECEIVE:
+        if (TestContext->IsServer) { // only server receives
+            TestContext->SendPhaseEndedEvent.Set();
+            TestContext->ExitCode = CancelOnLossContext::SuccessExitCode;
+        }
+        break;
+    case QUIC_STREAM_EVENT_PEER_SEND_ABORTED:
+        if (TestContext->IsServer) { // server-side 'cancel on loss' detection
+            TestContext->SendPhaseEndedEvent.Set();
+            TestContext->ExitCode = Event->PEER_SEND_ABORTED.ErrorCode;
+        }
+        else {
+            Status = QUIC_STATUS_INVALID_STATE;
+        }
+        break;
+    case QUIC_STREAM_EVENT_PEER_SEND_SHUTDOWN:
+        if (TestContext->IsServer) {
+            TestContext->SendPhaseEndedEvent.Set();
+        }
+        break;
+    case QUIC_STREAM_EVENT_SEND_COMPLETE:
+        if (!TestContext->IsServer) { // only client sends
+            if (!TestContext->IsDropScenario) { // if drop scenario, we use 'cancel on loss' event
+                TestContext->SendPhaseEndedEvent.Set();
+            }
+        }
+        else {
+            Status = QUIC_STATUS_INVALID_STATE;
+        }
+        break;
+    case QUIC_STREAM_EVENT_CANCEL_ON_LOSS:
+        if (!TestContext->IsServer && TestContext->IsDropScenario) { // only client sends & only happens if in drop scenario
+            Event->CANCEL_ON_LOSS.ErrorCode = CancelOnLossContext::ErrorExitCode;
+            TestContext->SendPhaseEndedEvent.Set();
+        }
+        else {
+            Status = QUIC_STATUS_INVALID_STATE;
+        }
+        break;
+    default:
+        break;
+    }
+
+    return Status;
+}
+
+_Function_class_(MsQuicConnectionCallback)
+QUIC_STATUS
+QuicCancelOnLossConnectionHandler(
+    _In_ struct MsQuicConnection* /* Connection */,
+    _In_opt_ void* Context,
+    _Inout_ QUIC_CONNECTION_EVENT* Event
+)
+{
+    if (Context == nullptr) {
+        return QUIC_STATUS_INVALID_PARAMETER;
+    }
+
+    auto TestContext = reinterpret_cast<CancelOnLossContext*>(Context);
+
+    QUIC_STATUS Status = QUIC_STATUS_SUCCESS;
+
+    switch (Event->Type) {
+    case QUIC_CONNECTION_EVENT_PEER_STREAM_STARTED:
+        TestContext->Stream = new MsQuicStream(
+            Event->PEER_STREAM_STARTED.Stream,
+            CleanUpManual,
+            QuicCancelOnLossStreamHandler,
+            Context);
+        break;
+    case QUIC_CONNECTION_EVENT_CONNECTED:
+        TestContext->ConnectedEvent.Set();
+        break;
+    default:
+        break;
+    }
+
+    return Status;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+_Function_class_(QUIC_LISTENER_CALLBACK)
+QUIC_STATUS
+QuicCancelOnLossListenerHandler(
```

Comment:
If you use the `MsQuicAutoAcceptListener` instead of `MsQuicListener` below, you can remove this.